bundles/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/protocol/SatelModule.java

```diff
@@ -0,0 +1,455 @@
+				timeoutTimer.start();
+				boolean sent = writeMessage(message);
+				timeoutTimer.stop();
+
+				if (sent) {
+					logger.trace("Waiting for response");
+					timeoutTimer.start();
+					response = this.readMessage();
+					timeoutTimer.stop();
+					if (response != null) {
+						logger.debug("Got response: {}", response);
+						command.handleResponse(response);
+					}
+				}
+
+				// if either send or receive failed, exit thread
+				if (!sent || response == null) {
+					break;
+				}
+
+			}
+		} catch (InterruptedException e) {
+			// exit thread
+		} catch (Exception e) {
+			// unexpected error, log and exit thread
+			logger.info("Unhandled exception occurred in communication loop, disconnecting.", e);
+		} finally {
+			// stop counting if thread interrupted
+			timeoutTimer.stop();
+		}
+
+		disconnect();
+	}
+
+	/*
+	 * Respawns communication thread in case on any error and interrupts it in
+	 * case read/write operations take too long.
+	 */
+	private class CommunicationWatchdog extends Timer implements TimeoutTimer {
+		private Thread thread;
+		private volatile long lastActivity;
+
+		public CommunicationWatchdog() {
+			this.thread = null;
+			this.lastActivity = 0;
+
+			this.schedule(new TimerTask() {
+				@Override
+				public void run() {
+					CommunicationWatchdog.this.checkThread();
+				}
+			}, 0, 1000);
+		}
+
+		@Override
+		public void start() {
+			this.lastActivity = System.currentTimeMillis();
+		}
+
+		@Override
+		public void stop() {
+			this.lastActivity = 0;
+		}
+
+		public void close() {
+			// cancel timer first to prevent reconnect
+			this.cancel();
+			// then stop communication thread
+			if (this.thread != null) {
+				this.thread.interrupt();
+				try {
+					this.thread.join();
+				} catch (InterruptedException e) {
+					// ignore
+				}
+			}
+		}
+
+		private void startCommunication() {
+			if (this.thread != null && this.thread.isAlive()) {
+				logger.error("Start communication canceled: communication thread is still alive");
+				return;
+			}
+			// start new thread
+			this.thread = new Thread(new Runnable() {
+				@Override
+				public void run() {
+					logger.debug("Communication thread started");
+					SatelModule.this.communicationLoop(CommunicationWatchdog.this);
+					logger.debug("Communication thread stopped");
+				}
+			});
+			this.thread.start();
+			// if module is not initialized yet, send version command
+			if (!SatelModule.this.isInitialized()) {
+				SatelModule.this.sendCommand(IntegraVersionCommand.buildMessage());
+			}
+		}
+
+		private void checkThread() {
+			if (this.thread != null && this.thread.isAlive()) {
+				long timePassed = (this.lastActivity == 0) ? 0 : System.currentTimeMillis() - this.lastActivity;
+
+				if (timePassed > SatelModule.this.timeout) {
+					logger.info("Send/receive timeout, disconnecting module.");
+					stop();
+					this.thread.interrupt();
+				}
+			} else {
+				startCommunication();
+			}
+		}
+	}
+}

```

Comment:
One thread is spawned to handle communication loop (get messages from the queue, send them to the module, receive response and generate events) and the second is a timer thread to implement timeouts, since reads and writes are blocking ones. If communication thread is hanging on a read for longer than "timeout" seconds it just kill the thread and reconnects to the module by spawning a new one.